File: apps/node/src/chapter15/prediction-api.ts
import { type Context, Hono } from "hono";
import { z } from "zod";
import { type Passenger, type Result, ModelNotFoundError } from "./domain.ts";
import type { PredictionService } from "./prediction-service.ts";

function notNegative(field: string) {
  return z
    .number({ error: `${field} は数値にしてください` })
    .nonnegative({ error: `${field} は 0 以上にしてください` });
}

function notNegativeInt(field: string) {
  return z
    .int({ error: `${field} は整数にしてください` })
    .nonnegative({ error: `${field} は 0 以上にしてください` });
}

function oneOf<
  const T extends readonly [string | number, ...(string | number)[]],
>(field: string, values: T) {
  return z.literal(values, {
    error: `${field} は ${values.join("、")} のどれかにしてください`,
  });
}

/** 映画の API の入力。JSON の項目名はドメインの Movie と同じ */
const movieRequestSchema = z.object({
  sns1: notNegative("sns1"),
  sns2: notNegative("sns2"),
  actor: notNegative("actor"),
  original: oneOf("original", [0, 1]),
});

/** 乗客の API の入力。年齢と乗船港は省略でき、欠損値の補完は第 8 章のパイプラインに任せる */
const passengerRequestSchema = z
  .object({
    pclass: oneOf("pclass", [1, 2, 3]),
    sex: oneOf("sex", ["male", "female"]),
    age: notNegative("age").nullish(),
    sib_sp: notNegativeInt("sib_sp"),
    parch: notNegativeInt("parch"),
    fare: notNegative("fare"),
    embarked: oneOf("embarked", ["C", "Q", "S"]).nullish(),
  })
  .transform((request): Passenger => ({
    pclass: request.pclass,
    sex: request.sex,
    age: request.age ?? null,
    sibSp: request.sib_sp,
    parch: request.parch,
    fare: request.fare,
    embarked: request.embarked ?? null,
  }));

/** 入力の検証結果。正しければ値を、不正なら理由の一覧を持つ */
type Validated<T> = { ok: true; value: T } | { ok: false; errors: string[] };

/** 本文を JSON として読み、スキーマで検証してドメインの型にする */
async function validate<T>(
  c: Context,
  schema: z.ZodType<T, unknown>,
): Promise<Validated<T>> {
  let body: unknown;
  try {
    body = await c.req.json();
  } catch {
    return { ok: false, errors: ["JSON の形式が正しくありません"] };
  }
  const parsed = schema.safeParse(body);
  return parsed.success
    ? { ok: true, value: parsed.data }
    : { ok: false, errors: parsed.error.issues.map((issue) => issue.message) };
}

/** 成功していれば値を返し、失敗していれば原因の例外を投げる */
function valueOrThrow<T>(result: Result<T>): T {
  if (!result.ok) {
    throw result.error;
  }
  return result.value;
}

export function createApp(service: PredictionService): Hono {
  const app = new Hono();

  app.post("/cinema/sales", async (c) => {
    const movie = await validate(c, movieRequestSchema);
    if (!movie.ok) {
      return c.json({ detail: movie.errors }, 422);
    }
    return c.json(valueOrThrow(service.predictSales(movie.value)));
  });

  app.post("/survived", async (c) => {
    const passenger = await validate(c, passengerRequestSchema);
    if (!passenger.ok) {
      return c.json({ detail: passenger.errors }, 422);
    }
    return c.json(valueOrThrow(service.predictSurvival(passenger.value)));
  });

  app.get("/health", (c) => {
    const models = service.health();
    const status = Object.values(models).every(Boolean) ? "ok" : "degraded";
    return c.json({ status, models });
  });

  // ドメインの例外を HTTP のステータスコードに変換する。内部の情報は応答に出さない
  app.onError((error, c) => {
    if (error instanceof ModelNotFoundError) {
      return c.json({ detail: error.message }, 503);
    }
    console.error(error);
    return c.json({ detail: "予測中にエラーが発生しました" }, 500);
  });

  return app;
}

File: apps/rust/tests/prediction_api.rs
//! 第 15 章の予測 API のテスト。置き場をスタブに差し替えるので、学習データが無くても走る。

use std::sync::Arc;

use axum::body::Body;
use axum::http::{Request, StatusCode};
use getting_started_ml::chapter15::{
    Error, ModelStore, Movie, Passenger, PredictionService, Result, SALES_MODEL, SURVIVAL_MODEL,
    SalesModel, SurvivalModel, api,
};
use tower::ServiceExt;

/// 置き場のスタブ。モデルがあるかどうかを差し替えられる。
struct StubStore {
    sales: bool,
    survival: bool,
}

/// 決まった値を返す興行収入のモデル。
struct FixedSales;

impl SalesModel for FixedSales {
    fn predict_sales(&self, _movie: Movie) -> Result<f64> {
        Ok(4321.5)
    }
}

/// 決まった値を返す生存予測のモデル。
struct FixedSurvival;

impl SurvivalModel for FixedSurvival {
    fn survives(&self, _passenger: &Passenger) -> Result<bool> {
        Ok(true)
    }
}

impl ModelStore for StubStore {
    fn load_sales_model(&self) -> Result<Box<dyn SalesModel>> {
        if self.sales {
            Ok(Box::new(FixedSales))
        } else {
            Err(Error::ModelNotFound(SALES_MODEL.to_string()))
        }
    }

    fn load_survival_model(&self) -> Result<Box<dyn SurvivalModel>> {
        if self.survival {
            Ok(Box::new(FixedSurvival))
        } else {
            Err(Error::ModelNotFound(SURVIVAL_MODEL.to_string()))
        }
    }
}

/// スタブの置き場を差し込んだ API を作る。
fn router(sales: bool, survival: bool) -> axum::Router {
    api::router(Arc::new(PredictionService::new(Box::new(StubStore {
        sales,
        survival,
    }))))
}

/// API を呼んで、状態コードと本文を返す。tower の `oneshot` でサーバーを起動せずに呼ぶ。
async fn call(
    method: &str,
    path: &str,
    body: &str,
    sales: bool,
    survival: bool,
) -> (StatusCode, String) {
    let request = Request::builder()
        .method(method)
        .uri(path)
        .header("content-type", "application/json")
        .body(Body::from(body.to_string()))
        .expect("要求を作れること");

    let response = router(sales, survival)
        .oneshot(request)
        .await
        .expect("応答が返ること");
    let status = response.status();
    let bytes = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .expect("本文を読めること");

    (
        status,
        String::from_utf8(bytes.to_vec()).expect("UTF-8 であること"),
    )
}

#[tokio::test]
async fn 興行収入を予測して_json_で返す() {
    let (status, body) = call(
        "POST",
        "/cinema/sales",
        r#"{"sns1": 100, "sns2": 2000, "actor": 300, "original": 1}"#,
        true,
        true,
    )
    .await;

    assert_eq!(status, StatusCode::OK);
    assert_eq!(body, r#"{"sales":4321.5}"#);
}

#[tokio::test]
async fn 生存を予測して_json_で返す() {
    let (status, body) = call(
        "POST",
        "/survived",
        r#"{"pclass": 1, "sex": "female", "age": 30, "sib_sp": 0, "parch": 0, "fare": 80, "embarked": "S"}"#,
        true,
        true,
    )
    .await;

    assert_eq!(status, StatusCode::OK);
    assert_eq!(body, r#"{"survived":true}"#);
}

#[tokio::test]
async fn 年齢と乗船港は省略できる() {
    let (status, _) = call(
        "POST",
        "/survived",
        r#"{"pclass": 3, "sex": "male", "sib_sp": 0, "parch": 0, "fare": 7.25}"#,
        true,
        true,
    )
    .await;

    assert_eq!(status, StatusCode::OK);
}

#[tokio::test]
async fn 入力が不正なら四百二十二と理由の一覧を返す() {
    let (status, body) = call("POST", "/cinema/sales", r#"{"sns1": 100}"#, true, true).await;

    assert_eq!(status, StatusCode::UNPROCESSABLE_ENTITY);
    assert!(body.contains("sns2 は必須です"), "本文 = {body}");
    assert!(body.contains("original は必須です"), "本文 = {body}");
}

#[tokio::test]
async fn 選択肢にない値を指摘する() {
    let (status, body) = call(
        "POST",
        "/survived",
        r#"{"pclass": 4, "sex": "unknown", "sib_sp": 0, "parch": 0, "fare": 10}"#,
        true,
        true,
    )
    .await;

    assert_eq!(status, StatusCode::UNPROCESSABLE_ENTITY);
    assert!(
        body.contains("pclass は 1、2、3 のどれかにしてください"),
        "本文 = {body}"
    );
    assert!(
        body.contains("sex は female、male のどれかにしてください"),
        "本文 = {body}"
    );
}

#[tokio::test]
async fn json_として読めなければ四百二十二を返し内部の型名を漏らさない() {
    for body in [r#"{"sns1": "たくさん"}"#, "{ここは JSON ではない", ""] {
        let (status, response) = call("POST", "/cinema/sales", body, true, true).await;

        assert_eq!(
            status,
            StatusCode::UNPROCESSABLE_ENTITY,
            "本文 = {response}"
        );
        assert!(
            response.contains("JSON の形式または値の型が正しくありません"),
            "本文 = {response}"
        );
        assert!(!response.contains("f64"), "本文 = {response}");
        assert!(!response.contains("MovieRequest"), "本文 = {response}");
    }
}

#[tokio::test]
async fn モデルが無ければ五百三を返す() {
    let (status, body) = call(
        "POST",
        "/cinema/sales",
        r#"{"sns1": 100, "sns2": 2000, "actor": 300, "original": 1}"#,
        false,
        true,
    )
    .await;

    assert_eq!(status, StatusCode::SERVICE_UNAVAILABLE);
    assert!(body.contains("モデルがありません: cinema"), "本文 = {body}");
}

#[tokio::test]
async fn ヘルスチェックはモデルごとの状態を返す() {
    let (status, body) = call("GET", "/health", "", true, true).await;

    assert_eq!(status, StatusCode::OK);
    assert_eq!(
        body,
        r#"{"status":"ok","models":{"cinema":true,"survived":true}}"#
    );

    let (_, degraded) = call("GET", "/health", "", true, false).await;

    assert!(
        degraded.contains(r#""status":"degraded""#),
        "本文 = {degraded}"
    );
}

#[tokio::test]
async fn 知らないパスは四百四許していないメソッドは四百五を返す() {
    let (status, _) = call("GET", "/unknown", "", true, true).await;
    assert_eq!(status, StatusCode::NOT_FOUND);

    let (status, _) = call("GET", "/cinema/sales", "", true, true).await;
    assert_eq!(status, StatusCode::METHOD_NOT_ALLOWED);
}
